File: epi_judge_python/is_tree_a_bst.py
from binary_tree_node import BinaryTreeNode
from test_framework import generic_test


def is_binary_tree_bst(tree: BinaryTreeNode) -> bool:
    # TODO - you fill in here.

    def is_binary_tree_bst(tree,lower_bound=float("-inf"),upper_bound=float("inf")):

        if tree is None:
            return True

        
        if not (lower_bound <= tree.data <= upper_bound):
            return False

        isBstL=is_binary_tree_bst(tree.left,lower_bound=lower_bound,upper_bound=tree.data)
        isBstR=is_binary_tree_bst(tree.right,lower_bound=tree.data,upper_bound=upper_bound)

        return isBstL and isBstR

    return is_binary_tree_bst(tree)


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('is_tree_a_bst.py', 'is_tree_a_bst.tsv',
                                       is_binary_tree_bst))

File: epi_judge_cpp/is_tree_a_bst.cc
#include <memory>

#include "binary_tree_node.h"
#include "test_framework/generic_test.h"
using std::unique_ptr;

bool IsBinaryTreeBSTBound(const unique_ptr<BinaryTreeNode<int>> &tree, int lowerbound, int upperbound)
{

  if (tree == nullptr)
  {
    return true;
  }

  if (!(lowerbound <= tree->data && tree->data <= upperbound))
  {
    return false;
  }

  bool lbool = IsBinaryTreeBSTBound(tree->left, lowerbound, tree->data);
  bool rbool = IsBinaryTreeBSTBound(tree->right, tree->data, upperbound);

  return lbool && rbool;
}

bool IsBinaryTreeBST(const unique_ptr<BinaryTreeNode<int>> &tree)
{
  // TODO - you fill in here.

  return IsBinaryTreeBSTBound(tree, std::numeric_limits<int>::min(), std::numeric_limits<int>::max());
}

int main(int argc, char *argv[])
{
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"tree"};
  return GenericTestMain(args, "is_tree_a_bst.cc", "is_tree_a_bst.tsv",
                         &IsBinaryTreeBST, DefaultComparator{}, param_names);
}
